libsolidity/formal/Predicate.cpp

```diff
@@ -0,0 +1,258 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsolidity/formal/Predicate.h>
+
+#include <libsolidity/formal/SMTEncoder.h>
+
+#include <libsolidity/ast/AST.h>
+
+#include <boost/algorithm/string/join.hpp>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::smtutil;
+using namespace solidity::frontend;
+using namespace solidity::frontend::smt;
+
+map<string, Predicate> Predicate::m_predicates;
+
+Predicate const* Predicate::create(
+	SortPointer _sort,
+	string _name,
+	EncodingContext& _context,
+	ASTNode const* _node
+)
+{
+	Predicate pred{_sort, _name, _context, _node};
+	string functorName = pred.functor().name;
+	solAssert(!m_predicates.count(functorName), "");
+	m_predicates.emplace(functorName, move(pred));
+	return &m_predicates.at(functorName);
+}
+
+Predicate::Predicate(
+	SortPointer _sort,
+	string _name,
+	EncodingContext& _context,
+	ASTNode const* _node
+):
+	m_predicate(_sort, move(_name), _context),
+	m_node(_node)
+{
+}
```

Comment:
So something like this maybe :-)? Together with deleting the move constructor?
```suggestion
Predicate const* Predicate::create(
	SortPointer _sort,
	string _name,
	EncodingContext& _context,
	ASTNode const* _node
)
{
	smt::SymbolicFunctionVariable predicate{_sort, move(_name), _context};
	string functorName = predicate.name;
	solAssert(!m_predicates.count(functorName), "");
	return &m_predicates.emplace(std::piecewise_construct, std::forward_as_tuple(functorName), std::forward_as_tuple(move(predicate), _node)).first->second;
}

Predicate::Predicate(
	smt::SymbolicFunctionVariable&& _predicate,
	ASTNode const* _node
):
	m_predicate(std::move(_predicate)),
	m_node(_node)
{
}
```

Reply:
lol let me try

Reply:
This worked, the only annoying thing is that now the actual constructor needs to be public, and I'm not sure there would be a way to guarantee that it's owned by the map. Thinking about it

Reply:
Ah yeah, hadn't thought about that... that's a bit annoying, that's true...

Reply:
There's no *good* solution to this... the go-to way if you *really* want it, is something like https://gist.github.com/valmat/69c67f94daf75452e2ae987aa67d55be , but that's kind of horrible...

Reply:
Not sure this is worth too much time - it'd be nicer if this ``Predicate`` *wasn't* movable, but it's actually just fine if it is, especially since the ``predicate(std::string const& _name);`` returns a ``const`` ref only anyways, so there won't be any public way to move anything out of the map anyways...

Reply:
The fact that this *would* work with a public constructor at least tells us that these *aren't* in fact moved around anywhere, so I'd be fine with moving back to (resp. staying with) the ``default``ed move constructor version.

Reply:
I think it's fine the way it is now, just need to keep in mind to never actually use that public constructor except in very local/special cases ;)